File: spec/julia/extra/MyCustomFx.jl
module MyCustomFx
export decode

function decode(key::Integer, flag::Bool, some_bytes::Vector{UInt8}, data::Vector{UInt8})
    _key = flag ? key : -key
    dst = Vector{UInt8}(undef, length(data))
    for i in eachindex(dst)
        dst[i] = (data[i] + _key)
    end
    dst
end
end

File: spec/java/src/io/kaitai/struct/testwrite/MyCustomFx.java
package io.kaitai.struct.testwrite;

import io.kaitai.struct.CustomProcessor;

public class MyCustomFx implements CustomProcessor {
    private int key;

    public MyCustomFx(int key, boolean flag, byte[] someBytes) {
        this.key = flag ? key : -key;
    }

    @Override
    public byte[] decode(byte[] src) {
        byte[] dst = new byte[src.length];
        for (int i = 0; i < src.length; i++)
            dst[i] = (byte) (src[i] + key);
        return dst;
    }

    @Override
    public byte[] encode(byte[] src) {
        int oldKey = key;
        key = -key;
        byte[] res = decode(src);
        key = oldKey;
        return res;
    }
}
